### PR TITLE
Debias & Splice Met products

### PR DESCRIPTION
**Pull request by crollinson**
Part of adding and updating new functionality in pecan to spatially and temporally downscale met data products.  In the PalEON met workflow, a debiasing step precedes the temporal downscaling that @jsimkins2 worked on. (Debugging and functionality improvements to that will come soon.). 

The PalEON workflow for generating long met ensembles that combine multiple products can be found here: https://github.com/PalEON-Project/modeling_met_ensemble (Yes, I know it's messy & the documentation is not that great.  I'll update it once the PEcAn end of things is ironed out.)

## Description
This pull requests adds two main functions for de-biasing (can also be seen as spatial downscaling) of met products: 
1. align.met -- takes two data products and matches their temporal resolution so that it can be then fed into downscaling code
2. debias.met.regression -- uses a training dataset such as LDAS or Ameriflux to sequentially adjust the mean climatology and seasonal cycle of multiple met variables, leveraging covariance among parameters where appropriate.

## Motivation and Context
These functions were added to allow coarse resolution data products to be adjusted to better match local conditions and/or allow multiple products to be combined to generate extended met drivers.  These were originally generated to allow the PalEON project to combine CMIP5 p1000 and historical runs, with modern reanalysis products such as CRUNCEP, NARR, and NLDAS.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) 
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: modules/data.atmosphere/R/align_met.R [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+# -----------------------------------
+# Parameters
+# -----------------------------------
+##' @param train.path - path to the dataset to be used to downscale the data
+##' @param source.path - data to be bias-corrected aligned with training data (from align.met)
+##' @param yrs.train - (optional) specify a specific years to be loaded for the training data; 
+##'                    prevents needing to load the entire dataset.  If NULL, all available years 
+##'                    will be loaded. If not null, should be a vector of numbers (so you can skip
+##'                    problematic years)
+##' @param n.ens  - number of ensemble members to generate and save
```
[the remark this answers is not in the document]
Comment (ankurdesai) [reply modules/data.atmosphere/R/align_met.R]: It would be good to eventually merge align.data and align.met in a single functions that aligns any two time series. But understand the limitation on that for now.

File: modules/data.atmosphere/R/debias_met_regression.R [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+# Begin Function
+#----------------------------------------------------------------------
+
+
+debias.met.regression <- function(train.data, source.data, n.ens, vars.debias=NULL, CRUNCEP=FALSE,
+                                  pair.anoms = TRUE, pair.ens = FALSE, uncert.prop="mean", resids = FALSE, seed=Sys.Date(),
+                                  outfolder, yrs.save=NULL, ens.name, ens.mems=NULL, lat.in, lon.in,
+                                  save.diagnostics=TRUE, path.diagnostics=NULL,
+                                  parallel = FALSE, n.cores = NULL, overwrite = TRUE, verbose = FALSE) {
+  library(MASS)
```
Comment (ankurdesai) [inline modules/data.atmosphere/R/debias_met_regression.R]: Same comment as before - on moving library calls into export, and explicitly labelling with ::

File: modules/data.atmosphere/R/debias_met_regression.R [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+        k=max(k, 4) # we can't have less than 4 knots
+        
+        # plot(anom.train ~ anom.raw, data=dat.anom)
+        # abline(a=0, b=1, col="blue")
+        # abline(lm(anom.train ~ anom.raw, data=dat.anom), col="red", lty="dashed")
+        
+        # Modeling in the predicted value from mod.bias
+        dat.anom$pred <- predict(mod.bias, newdata=dat.anom)
+        
+        if (v %in% c("air_temperature", "air_temperature_maximum", "air_temperature_minimum")){
```
Comment (ankurdesai) [inline modules/data.atmosphere/R/debias_met_regression.R]: This if statement with many switches is hard to follow. Would swtich statement be easier to read?

File: modules/data.atmosphere/R/extract_local_CMIP5.R [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    ntime = nday # leap year or not; time slice (coerce to daily)
+    
+    loc.file <- file.path(outfolder, paste(model, scenario, ensemble_member, str_pad(y.now, width=4, side="left",  pad="0"), "nc", sep = "."))
+    
+    
+    ## Create dimensions
+    dim.lat <- ncdim_def(name='latitude', units='degree_north', vals=lat.in, create_dimvar=TRUE)
+    dim.lon <- ncdim_def(name='longitude', units='degree_east', vals=lon.in, create_dimvar=TRUE)
+    dim.time <- ncdim_def(name='time', units="sec", vals=seq((min(days.use)+1-1/24)*24*360, (max(days.use)+1-1/24)*24*360, length.out=ntime), create_dimvar=TRUE, unlim=TRUE)
```
Comment (jsimkins2) [inline modules/data.atmosphere/R/extract_local_CMIP5.R]: Not sure what CF time units are for historical GCMs, but future GCMs time should be "days since YYYY".  Just so we are keeping consistent with the direction we want to move in for dimensions, I think we should consider editing this. Here is an example from download.CRUNCEP.

days_elapsed <- (1:ntime) * 6/24 - 3/24 # data are 6-hourly, with timestamp at center of interval
time <- ncdf4::ncdim_def(name = "time", units = paste0("days since ", year, "-01-01T00:00:00Z"),
                             vals = as.array(days_elapsed), create_dimvar = TRUE, unlim = TRUE)

File: modules/data.atmosphere/R/extract_local_CMIP5.R [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+##' @export
+##' @examples
+# -----------------------------------
+extract.local.CMIP5 <- function(outfolder, in.path, start_date, end_date, site_id, lat.in, lon.in, 
+                                model , scenario , ensemble_member = "r1i1p1",
+                                overwrite = FALSE, verbose = FALSE, ...){
+  library(lubridate)
+  library(ncdf4)
+  library(stringr)
+
```
Comment (jsimkins2) [inline modules/data.atmosphere/R/extract_local_CMIP5.R]: Library calls shouldn't be needed if each function from these packages is considered throughout (i.e. ncdf4::ncvar_get)

File: modules/data.atmosphere/R/extract_local_CMIP5.R [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  
+  # Figuring out what we have daily for and what we only have monthly for
+  vars.gcm.day <- dir(file.path(in.path, "day"))
+  vars.gcm.mo <- dir(file.path(in.path, "month"))
+  vars.gcm.mo <- vars.gcm.mo[!vars.gcm.mo %in% vars.gcm.day]
+  
+  vars.gcm <- c(vars.gcm.day, vars.gcm.mo)
+  
+  # Rewriting the dap name to get the closest variable that we have for the GCM (some only give uss stuff at sea level)
+  library(car) # having trouble gettins stuff to work otherwise
```
Comment (jsimkins2) [inline modules/data.atmosphere/R/extract_local_CMIP5.R]: same comment as the library() call above, standard PEcAn convention is using car::function instead of calling the entire library.

File: modules/data.atmosphere/R/align_met.R [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+      
+      # Create a data frame with all the important time info
+      # center the hour step
+      df.time <- data.frame(Year=yr.now, DOY=rep(1:nday, each=1/step.day), Hour=rep(stamps.hr, ntime))
+      df.time$Date <- strptime(paste(df.time$Year, df.time$DOY, df.time$Hour, sep="-"), format=("%Y-%j-%H"), tz="UTC")
+      met.out$dat.train[["time"]] <- rbind(met.out$dat.train$time, df.time)
+      
+      # Extract the met info, making matrices with the appropriate number of ensemble members
+      for(v in names(ncT$var)){
+        df.tem <- matrix(rep(ncvar_get(ncT, v), n.trn), ncol=n.trn, byrow=F)
```
[the remark this answers is not in the document]
Comment (crollinson) [reply modules/data.atmosphere/R/align_met.R]: ncvar_get can only get one variable at a time.  All the stuff with n.trn may not be necessary in this instance since I did some restructuring, but we do need to make sure it ends up as a matrix rather than a vector.

File: modules/data.atmosphere/R/align_met.R [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+      nday <- ifelse(leap_year(yr.now), 366, 365)
+      ntime <- length(ncT$dim$time$vals)
+      step.day <- nday/ntime
+      step.hr  <- step.day*24
+      stamps.hr <- seq(step.hr/2, by=step.hr, length.out=1/step.day) # Time stamps centered on period
+      
+      # Create a data frame with all the important time info
+      # center the hour step
+      df.time <- data.frame(Year=yr.now, DOY=rep(1:nday, each=1/step.day), Hour=rep(stamps.hr, ntime))
+      df.time$Date <- strptime(paste(df.time$Year, df.time$DOY, df.time$Hour, sep="-"), format=("%Y-%j-%H"), tz="UTC")
```
[the remark this answers is not in the document]
Comment (crollinson) [reply modules/data.atmosphere/R/align_met.R]: there may very well be a better way to do it, but to get this working, I'll ask that someone else figure it out while I move on to debugging the temporal downscaling.

File: modules/data.atmosphere/R/align_met.R [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    yrs.file <- matrix(unlist(yrs.file), ncol=length(yrs.file[[1]]), byrow=T)
+    yrs.file <- as.numeric(yrs.file[,ncol(yrs.file)-1]) # Assumes year is always last thing before the file extension
+    
+    if(!is.null(yrs.train)){
+      files.train <- files.train[which(yrs.file %in% yrs.train)]
+      yrs.file <- yrs.file[which(yrs.file %in% yrs.train)]
+    }
+    
+    # Loop through the .nc files putting everything into a list
+    print("Processing Training Data")
```
[the remark this answers is not in the document]
Comment (crollinson) [reply modules/data.atmosphere/R/align_met.R]: I am not familiar with the PEcAn.logger package, so I'm going to ask one of the core PEcAn team members to do the appropriate code fixes so I can move onto debugging temporal downscaling.

File: modules/data.atmosphere/R/debias_met_regression.R (hunk 3 above)
Comment (crollinson) [reply modules/data.atmosphere/R/debias_met_regression.R]: I agree this is cumbersome and hard to follow, but am not familiar with switch syntax.  If someone can point me to a good example to follow, I can play around with it if I get a chance later.

File: modules/data.atmosphere/R/extract_local_CMIP5.R (hunk 4 above)
Comment (crollinson) [reply modules/data.atmosphere/R/extract_local_CMIP5.R]: I think I found the past GCMs to be inconsistent and wasn't sure if PEcAn had a standard reference time that would make sense for our p1000 dates.

File: modules/data.atmosphere/R/align_met.R [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+# 1. This assumes that both the trian and source data are in *at least* daily resolution
+# -----------------------------------
+
+#----------------------------------------------------------------------
+# Begin Function
+#----------------------------------------------------------------------
+align.met <- function(train.path, source.path, yrs.train=NULL, n.ens=NULL, pair.mems = TRUE, seed=Sys.Date(), verbose = FALSE) {
+  # Load required libraries
+  library(ncdf4)
+  library(lubridate)
```
[the remark this answers is not in the document]
Comment (crollinson) [reply modules/data.atmosphere/R/align_met.R]: I started to put the references in the calls, but didn't do a full re-run, so I haven't removed the library calls yet

File: modules/data.atmosphere/R/align_met.R (hunk 9 above)
[the remark this answers is not in the document]
Comment (ankurdesai) [reply modules/data.atmosphere/R/align_met.R]: Basically just replace print with logger.info("...") , unless it's a warning, then use logger.warn, logger.error to throw an error logger.severe to force execution of pecan workflow to stop.
